### PR TITLE
Issue/#91 re-structure tests

### PR DESCRIPTION
**Pull request by dma1979**
#### Some additional points:
1) Currently latest JUnit 4 Generation is used - 4.12 .
What do you think about using 5G(jupiter)?
It will allow to use more functional approach(lambdas) even for testing itself. Ane even more: extensions, conditions, parameterized tests, dynamic tests and etc.

2) What if we apply grouping: categories in JUnit 4G or tags in JUnit 5G?
F.e. we can split all test to 2 categories/tags: 
- api
- internal

From the [documentation](https://junit.org/junit5/docs/5.4.2/api/org/junit/jupiter/api/Tag.html) :
tags are used to filter which tests are executed for a given test plan. For example, a development team may tag tests with values such as "fast", "slow", "ci-server", etc. and then supply a list of tags to be used for the current test plan, potentially dependent on the current environment.

### CODE REVIEW

**Comment by amaembo** [comment]
Please resolve conflicts.

**Comment by dma1979** [comment]
Final improvements for StreamEx tests: 
- 477 tests were split to 2 categories
  -  399 API tests 
   - 78 internal tests
 - native junit assertThrows(since 4.13) is used

**Comment by dma1979** [comment]
Tagir, could you please give your feedback?

File: src/test/java/one/util/streamex/api/DoubleCollectorTest.java [hunk 1]
```diff
@@ -1,174 +1,181 @@
...
+    public void testAdaptor() {
+        assertEquals(499.5, IntStreamEx.range(1000).asDoubleStream().collect(
+            DoubleCollector.of(DoubleCollector.summarizing())).getAverage(), 0.000001);
+        assertEquals(499.5, IntStreamEx.range(1000).parallel().asDoubleStream().collect(
+            DoubleCollector.of(Collectors.summarizingDouble(Double::doubleValue))).getAverage(), 0.000001);
+    }
+
+    @Test
+    public void testReducing() {
+        Assert.assertEquals(7.0, DoubleStreamEx.of(1.0, 2.0, 3.5).collect(DoubleCollector.reducing(1.0, (a, b) -> a * b)), 0.0);
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/DoubleCollectorTest.java:151]: Why explicit `Assert.` qualifier was added here?

File: src/test/java/one/util/streamex/api/DoubleCollectorTest.java [hunk 2]
```diff
@@ -1,174 +1,181 @@
...
+        assertEquals(7.0, DoubleStreamEx.of(1.0, 2.0, 3.5).parallel().collect(
+            DoubleCollector.reducing(1.0, (a, b) -> a * b)), 0.0);
+    }
+
+    @Test
+    public void testMapping() {
+        assertEquals(10100.0, IntStreamEx.rangeClosed(0, 100).asDoubleStream().collect(
+            DoubleCollector.mapping(x -> x * 2, DoubleCollector.summing())), 0.0);
+
+        Assert.assertArrayEquals(LongStreamEx.of(1, 1, 2, 3).toArray(), DoubleStreamEx.of(0.8, 1.3, 1.7, 2.9).collect(
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/DoubleCollectorTest.java:161]: Same

File: src/test/java/one/util/streamex/api/DoubleStreamExTest.java [hunk 3]
```diff
@@ -1,537 +1,543 @@
...
+    public void testBasics() {
+        assertFalse(DoubleStreamEx.of(1).isParallel());
+        assertTrue(DoubleStreamEx.of(1).parallel().isParallel());
+        assertFalse(DoubleStreamEx.of(1).parallel().sequential().isParallel());
+        AtomicInteger i = new AtomicInteger();
+        try (DoubleStreamEx s = DoubleStreamEx.of(1).onClose(i::incrementAndGet)) {
+            assertEquals(1, s.count());
+        }
+        assertEquals(1, i.get());
+        Assert.assertEquals(6, IntStreamEx.range(0, 4).asDoubleStream().sum(), 0);
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/DoubleStreamExTest.java:140]: Same.

File: src/test/java/one/util/streamex/api/EmitterTest.java [hunk 4]
```diff
@@ -1,181 +1,191 @@
...
+        assertEquals(asList(1, 1, 2, 3, 5, 8, 13, 21, 34, 55), fibonacci().map(BigInteger::intValueExact).limit(10)
+                .toList());
+
+        assertEquals(asList("aa", "aabbb", "aabbbc"), scanLeft(asList("aa", "bbb", "c").iterator(), "", String::concat)
+                .stream().toList());
+        
+        assertEquals(asList(4, 4, 4, 4, 3, 3, 3, 2, 2, 1), flatTest(4).stream().toList());
+        assertEquals(asList(4, 4, 4, 4, 3, 3, 3, 2, 2), flatTest(4).stream().limit(9).toList());
+
+        TestHelpers.checkSpliterator("flatTest", asList(4, 4, 4, 4, 3, 3, 3, 2, 2, 1),
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/EmitterTest.java:180]: Explicit `TestHelpers.` qualifiers.

File: src/test/java/one/util/streamex/api/EntryStreamTest.java [hunk 5]
```diff
@@ -86,10 +87,11 @@ private static void checkAsString(String expected, EntryStream<?, ?> stream) {
     @Test
     public void testCreate() {
         assertEquals(0, EntryStream.empty().count());
-        assertEquals(0, EntryStream.empty().count());
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/EntryStreamTest.java:89]: The duplication here was intentional and it actually caught a bug in the early version of StreamEx when `EntryStream.emtpy()` returned a shared object. Please restore it.

File: src/test/java/one/util/streamex/api/IntCollectorTest.java [hunk 6]
```diff
@@ -1,245 +1,250 @@
...
+
+/**
+ * @author Tagir Valeev
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class IntCollectorTest {
+    @Test
+    public void testJoining() {
+        String expected = IntStream.range(0, 10000).mapToObj(String::valueOf).collect(Collectors.joining(", "));
+        Assert.assertEquals(expected, IntStreamEx.range(10000).collect(IntCollector.joining(", ")));
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/IntCollectorTest.java:49]: Explicit qualifier.

File: src/test/java/one/util/streamex/api/IntCollectorTest.java [hunk 7]
```diff
@@ -1,245 +1,250 @@
...
+    @Test
+    public void testAdaptor() {
+        assertEquals(499500, (int) IntStreamEx.range(0, 1000).collect(IntCollector.of(IntCollector.summing())));
+        assertEquals(499500, (int) IntStreamEx.range(0, 1000).collect(
+            IntCollector.of(Collectors.summingInt(Integer::intValue))));
+    }
+
+    @Test
+    public void testMapping() {
+        Assert.assertArrayEquals(IntStreamEx.range(1000).asDoubleStream().toArray(), IntStreamEx.range(1000).collect(
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/IntCollectorTest.java:229]: Same

File: src/test/java/one/util/streamex/api/IntStreamExTest.java [hunk 8]
```diff
@@ -1,802 +1,806 @@
...
+/**
+ * @author Tagir Valeev
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class IntStreamExTest {
+    private static final byte[] EVEN_BYTES = new byte[] { 2, 4, 6, 8, 10 };
+
+    @Test
+    public void testCreate() {
+        Assert.assertArrayEquals(new int[] {}, IntStreamEx.empty().toArray());
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/IntStreamExTest.java:78]: Same

File: src/test/java/one/util/streamex/api/LongStreamExTest.java [hunk 9]
```diff
@@ -1,632 +1,637 @@
...
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class LongStreamExTest {
+    final LongConsumer EMPTY = l -> {
+        // nothing
+    };
+
+    @Test
+    public void testCreate() {
+        Assert.assertArrayEquals(new long[] {}, LongStreamEx.empty().toArray());
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/LongStreamExTest.java:75]: Explicit qualifier

File: src/test/java/one/util/streamex/api/LongStreamExTest.java [hunk 10]
```diff
@@ -1,632 +1,637 @@
...
+                s -> assertEquals(OptionalLong.of(5), s.prefix(Long::sum).findFirst(x -> x > 4)));
+        longStreamEx(() -> LongStreamEx.constant(1, 5),
+                s -> assertEquals(OptionalLong.empty(), s.prefix(Long::sum).findFirst(x -> x > 6)));
+    }
+
+    @Test
+    public void testIntersperse() {
+        assertArrayEquals(new long[] { 1, 0, 10, 0, 100, 0, 1000 }, LongStreamEx.of(1, 10, 100, 1000).intersperse(0)
+                .toArray());
+        Assert.assertEquals(0L, IntStreamEx.empty().intersperse(1).count());
```
Comment (amaembo) [inline src/test/java/one/util/streamex/api/LongStreamExTest.java:635]: Same

**Review by amaembo — CHANGES_REQUESTED** [review]
Seems some qualifiers were added due to move refactoring...

**Pull request merged**